pkg/icingadb/db.go

```diff
@@ -296,6 +311,49 @@ func (db *DB) NamedBulkExec(
 	return g.Wait()
 }
 
+// NamedExecContextBulk performs db.NamedExecContext(ctx, query, rows).
+// If that fails due to upserting the same row twice (Postgres only), it splits rows and upserts them in a transaction.
+func (db *DB) NamedExecContextBulk(ctx context.Context, query string, rows []contracts.Entity) error {
+	_, err := db.NamedExecContext(ctx, query, rows)
+	if err != nil && len(rows) > 1 && isDuplicateRowInUpsert(err) {
+		tx, err := db.BeginTxx(ctx, nil)
+		if err != nil {
+			return errors.Wrap(err, "can't start transaction")
+		}
+
+		chunk := rows[:len(rows)/2]
+
+		for {
+			if _, err := tx.NamedExecContext(ctx, query, chunk); err != nil {
+				if len(chunk) > 1 && isDuplicateRowInUpsert(err) {
+					chunk = chunk[:len(chunk)/2]
+					continue
+				}
+
+				return errors.Wrap(err, "can't execute statement in transaction")
+			}
+
+			if chunk = rows[len(chunk):]; len(chunk) > 0 {
+				continue
+			}
+
+			break
+		}
+
+		if err := tx.Commit(); err != nil {
+			return errors.Wrap(err, "can't commit transaction")
+		}
+	}
+
+	return err
+}
+
+func isDuplicateRowInUpsert(err error) bool {
+	var pe *pq.Error
+	return errors.As(err, &pe) && pe.Code == "21000" &&
+		pe.Message == "ON CONFLICT DO UPDATE command cannot affect row a second time"
```

Comment:
Does this really need to check the error message here? Also can that message change based on server locale or something like that?

Reply:
1. Yes, 21000 is a little ambiguous.
2. I have even trouble with localizing the Postgres shell:

```
postgres@parallels-Parallels-Virtual-Platform:~$ LC_ALL=ru_RU psql
psql (12.2 (Ubuntu 12.2-4))
Type "help" for help.

postgres=# 
```